homeassistant/components/trafikverket_train/coordinator.py

```diff
@@ -0,0 +1,132 @@
+    days_ahead = weekday - fromdate.weekday()
+    if days_ahead <= 0:
+        days_ahead += 7
+    return fromdate + timedelta(days_ahead)
+
+
+def next_departuredate(departure: list[str]) -> date:
+    """Calculate the next departuredate from an array input of short days."""
+    today_date = date.today()
+    today_weekday = date.weekday(today_date)
+    if WEEKDAYS[today_weekday] in departure:
+        return today_date
+    for day in departure:
+        next_departure = WEEKDAYS.index(day)
+        if next_departure > today_weekday:
+            return next_weekday(today_date, next_departure)
+    return next_weekday(today_date, WEEKDAYS.index(departure[0]))
+
+
+class TVDataUpdateCoordinator(DataUpdateCoordinator):
+    """A Trafikverket Data Update Coordinator."""
+
+    def __init__(
+        self,
+        hass: HomeAssistant,
+        entry: ConfigEntry,
+        to_station: StationInfo,
+        from_station: StationInfo,
+    ) -> None:
+        """Initialize the Trafikverket coordinator."""
+        super().__init__(
+            hass,
+            _LOGGER,
+            name=DOMAIN,
+            update_interval=TIME_BETWEEN_UPDATES,
+        )
+        self._train_api = TrafikverketTrain(
+            async_get_clientsession(hass), entry.data[CONF_API_KEY]
+        )
+        self.from_station: StationInfo = from_station
+        self.to_station: StationInfo = to_station
+        self._time: time | None = dt_util.parse_time(entry.data[CONF_TIME])
+        self._weekdays: list[str] = entry.data[CONF_WEEKDAY]
+
+    async def _async_update_data(self) -> dict[str, StateType | datetime | timedelta]:
+        """Fetch data from Trafikverket."""
+
+        when = dt_util.now()
+        state: TrainStop | None = None
+        if self._time:
+            departure_day = next_departuredate(self._weekdays)
+            when = datetime.combine(
+                departure_day,
+                self._time,
+                dt_util.get_time_zone(self.hass.config.time_zone),
+            )
+        try:
+            if self._time:
+                state = await self._train_api.async_get_train_stop(
+                    self.from_station, self.to_station, when
+                )
+            else:
+                state = await self._train_api.async_get_next_train_stop(
+                    self.from_station, self.to_station, when
+                )
+        except ValueError as error:
+            if "Invalid authentication" in error.args[0]:
+                raise ConfigEntryAuthFailed from error
+            raise UpdateFailed(
+                f"Train departure {when} encountered a problem: {error}"
+            ) from error
+
+        departure_time = state.advertised_time_at_location
+        if state.time_at_location:
+            departure_time = state.time_at_location
+        if state.estimated_time_at_location:
+            departure_time = state.estimated_time_at_location
```

Comment:
This logic means the latter wins if it is true, so can be written as:

```suggestion
        departure_time = state.advertised_time_at_location
        if state.estimated_time_at_location:
            departure_time = state.estimated_time_at_location
        elif state.time_at_location:
            departure_time = state.time_at_location
```